File: javascript/src/core/column.js
import { FlyqlError } from './exceptions.js'
import { Type, parseFlyQLType } from '../flyql_type.js'

/**
 * The canonical, schema-aware Column used by the validator. Dialect
 * generators have their own opaque Column classes; bridge with the
 * dialect's `toFlyQLSchema` helper to feed the validator.
 *
 * `JSONString` is an orthogonal capability flag — see Tech Decision #5.
 * It is NOT validated against `type`.
 */
export class Column {
    constructor(
        name,
        jsonString,
        type,
        {
            values = [],
            displayName = '',
            rawIdentifier = '',
            matchName = null,
            suggest = true,
            children = null,
            autocomplete,
        } = {},
    ) {
        this.name = name
        this.jsonString = jsonString
        this.type = type
        this.values = values
        this.displayName = displayName
        this.rawIdentifier = rawIdentifier
        this.matchName = matchName !== null ? matchName : name
        this.suggest = suggest
        this.children = children
        if (autocomplete !== undefined) this.autocomplete = autocomplete
    }

    get isNested() {
        return this.children != null
    }

    withRawIdentifier(identifier) {
        this.rawIdentifier = identifier
        return this
    }
}

export class ColumnSchema {
    constructor(columns) {
        this._columns = columns
        this._byLowerName = {}
        for (const [k, v] of Object.entries(columns)) {
            this._byLowerName[k.toLowerCase()] = v
            if (v != null && v.children != null) {
                _lowercaseChildren(v)
            }
        }
    }

    get columns() {
        return this._columns
    }

    get(name) {
        return this._byLowerName[name.toLowerCase()] || null
    }

    resolve(segments) {
        if (!segments || segments.length === 0) return null
        let col = this._byLowerName[segments[0].toLowerCase()]
        if (col == null) return null
        for (let i = 1; i < segments.length; i++) {
            if (col.children == null) return null
            col = col.children[segments[i].toLowerCase()]
            if (col == null) return null
        }
        return col
    }

    resolveWithPath(segments) {
        if (!segments || segments.length === 0) return null
        const rootKey = _findOriginalKey(this._columns, segments[0])
        if (!rootKey) return null
        let col = this._columns[rootKey]
        if (col == null) return null
        let parentPath = rootKey

        for (let i = 1; i < segments.length; i++) {
            if (col.children == null) return null
            const childCol = col.children[segments[i].toLowerCase()]
            if (childCol == null) return null
            parentPath += '.' + (childCol.name || segments[i])
            col = childCol
        }
        return { column: col, parentPath }
    }

    /**
     * Build a ColumnSchema from a `{name: {type, children, suggest, values, jsonstring}}`
     * dict. Strict mode: an unknown `type` value throws FlyqlError. The
     * legacy key `normalized_type` is detected and throws a targeted
     * migration error.
     */
    static fromPlainObject(obj) {
        const columns = {}
        for (const [name, raw] of Object.entries(obj)) {
            const col = _columnFromPlainObject(name, raw)
            if (col != null) columns[name] = col
        }
        return new ColumnSchema(columns)
    }

    static fromColumns(columns) {
        const m = {}
        for (const col of columns) {
            if (!(col.matchName in m)) {
                m[col.matchName] = col
            }
        }
        return new ColumnSchema(m)
    }
}

function _findOriginalKey(obj, target) {
    const lower = target.toLowerCase()
    for (const key of Object.keys(obj)) {
        if (key.toLowerCase() === lower) return key
    }
    return null
}

function _lowercaseChildren(col) {
    if (col.children == null) return
    const lowered = {}
    for (const [k, child] of Object.entries(col.children)) {
        lowered[k.toLowerCase()] = child
        if (child != null && child.children != null) {
            _lowercaseChildren(child)
        }
    }
    col.children = lowered
}

function _columnFromPlainObject(name, raw) {
    if (raw == null || typeof raw !== 'object') return null
    if ('normalized_type' in raw || 'normalizedType' in raw) {
        throw new FlyqlError(
            `column "${name}": "normalized_type"/"normalizedType" field has been ` +
                `renamed to "type" in canonical column JSON; see migration guide at ` +
                `docs.flyql.dev/advanced/column-types`,
        )
    }
    let children = null
    if (raw.children != null && typeof raw.children === 'object') {
        children = {}
        for (const [childName, childRaw] of Object.entries(raw.children)) {
            const child = _columnFromPlainObject(childName, childRaw)
            if (child != null) children[childName] = child
        }
    }
    // Lenient: unknown type strings are coerced to Type.Unknown so
    // editor-style raw strings (e.g. 'enum') flow through without
    // throwing; the editor normalizer remaps them at a later stage.
    // The legacy `normalized_type` key is still a hard error (above).
    const typeStr = raw.type || ''
    let flyqlType = Type.Unknown
    if (typeStr) {
        try {
            flyqlType = parseFlyQLType(typeStr)
        } catch (_) {
            flyqlType = typeStr // preserve raw string for the editor normalizer
        }
    }
    return new Column(name, !!raw.jsonstring, flyqlType, {
        values: raw.values || [],
        suggest: raw.suggest !== undefined ? raw.suggest : true,
        matchName: name,
        children,
        autocomplete: raw.autocomplete,
        displayName: raw.display_name || raw.displayName || '',
    })
}

File: python/flyql/core/column.py
"""Canonical Column and ColumnSchema used by the validator.

Dialect-specific generators have their own opaque Column classes; use the
dialect's ``to_flyql_schema`` helper to bridge from a dialect column list
to a :class:`ColumnSchema`.
"""

from typing import Any, Dict, List, Optional

from flyql.core.exceptions import FlyqlError
from flyql.flyql_type import Type, parse_flyql_type


class Column:
    """The canonical, schema-aware column type used by the validator.

    ``column_type`` (NOT ``type``) is the parameter name because ``type``
    is a Python builtin and shadowing it inside ``__init__`` would
    generate pylint warnings and risk silent bugs. See Tech Decision #15.
    """

    def __init__(
        self,
        name: str,
        jsonstring: bool,
        column_type: Type,
        values: Optional[List[str]] = None,
        display_name: str = "",
        raw_identifier: str = "",
        match_name: Optional[str] = None,
        suggest: bool = True,
        children: Optional[Dict[str, "Column"]] = None,
    ) -> None:
        self.name = name
        # JSONString is an orthogonal capability flag — see Tech Decision #5.
        # NOT validated against ``column_type``.
        self.jsonstring = jsonstring
        self.type: Type = column_type
        self.values = values or []
        self.display_name = display_name
        self.raw_identifier = raw_identifier
        self.match_name = match_name if match_name is not None else name
        self.suggest = suggest
        self.children = children

    @property
    def is_nested(self) -> bool:
        return self.children is not None

    def with_raw_identifier(self, identifier: str) -> "Column":
        self.raw_identifier = identifier
        return self


class ColumnSchema:
    """Wraps a set of columns with case-insensitive lookup and nested path resolution."""

    def __init__(self, columns: Dict[str, "Column"]) -> None:
        self._columns = columns
        self._by_lower_name: Dict[str, Column] = {}
        for k, v in columns.items():
            self._by_lower_name[k.lower()] = v
            if v is not None and v.children is not None:
                _lowercase_children(v)

    @property
    def columns(self) -> Dict[str, "Column"]:
        return self._columns

    def get(self, name: str) -> Optional["Column"]:
        return self._by_lower_name.get(name.lower())

    def resolve(self, segments: List[str]) -> Optional["Column"]:
        """Walk nested column tree by segments (case-insensitive).
        Returns None if any segment is unresolvable."""
        if not segments:
            return None
        col = self._by_lower_name.get(segments[0].lower())
        if col is None:
            return None
        for seg in segments[1:]:
            if col.children is None:
                return None
            col = col.children.get(seg.lower())
            if col is None:
                return None
        return col

    @classmethod
    def from_columns(cls, columns: List["Column"]) -> "ColumnSchema":
        """Build a ColumnSchema from a flat Column list, keyed by match_name."""
        m: Dict[str, Column] = {}
        for c in reversed(columns):
            m[c.match_name] = c
        return cls(m)

    @classmethod
    def from_plain_object(cls, obj: Dict[str, Any]) -> "ColumnSchema":
        """Recursively convert ``{name: {type, children, suggest, values, jsonstring}}``
        dicts.

        Strict mode: an unknown ``type`` value raises :class:`FlyqlError`.
        The legacy key ``normalized_type`` is detected and raises a
        targeted migration error pointing at
        ``docs.flyql.dev/advanced/column-types``.
        """
        m: Dict[str, Column] = {}
        for name, raw in obj.items():
            col = _column_from_plain_object(name, raw)
            if col is not None:
                m[name] = col
        return cls(m)


def _lowercase_children(col: Column) -> None:
    if col.children is None:
        return
    lowered: Dict[str, Column] = {}
    for k, child in col.children.items():
        lowered[k.lower()] = child
        if child is not None and child.children is not None:
            _lowercase_children(child)
    col.children = lowered


def _column_from_plain_object(name: str, raw: Any) -> Optional[Column]:
    if not isinstance(raw, dict):
        return None
    if "normalized_type" in raw:
        raise FlyqlError(
            f"column {name!r}: 'normalized_type' field has been renamed to 'type' "
            "in canonical column JSON; see migration guide at "
            "docs.flyql.dev/advanced/column-types"
        )
    children: Optional[Dict[str, Column]] = None
    if "children" in raw and isinstance(raw["children"], dict):
        children = {}
        for child_name, child_raw in raw["children"].items():
            child = _column_from_plain_object(child_name, child_raw)
            if child is not None:
                children[child_name] = child
    type_str = raw.get("type", "")
    column_type = parse_flyql_type(type_str) if type_str else Type.Unknown
    return Column(
        name=name,
        jsonstring=bool(raw.get("jsonstring", False)),
        column_type=column_type,
        values=raw.get("values", []),
        suggest=raw.get("suggest", True),
        match_name=name,
        children=children,
    )
